plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,198 @@
+#include "wayfire/core.hpp"
+
+#include "wayfire/seat.hpp"
+#include "wayfire/scene-input.hpp"
+#include "wayfire/scene-operations.hpp"
+#include "wayfire/unstable/wlr-surface-node.hpp"
+#include "wayfire/output.hpp"
+#include "wayfire/output-layout.hpp"
+
+#include <wayfire/plugin.hpp>
+#include <wayfire/nonstd/wlroots-full.hpp>
+#include <wayfire/util/log.hpp>
+
+class wf_session_lock_impl : public wf::plugin_interface_t
+{
+    class lock_surface_node_t : public wf::scene::wlr_surface_node_t
+    {
+      public:
+        lock_surface_node_t(wlr_surface *surface, bool autocommit) :
+            wf::scene::wlr_surface_node_t(surface, autocommit)
+        {
+            interaction = new lock_surface_keyboard_interaction(surface);
+        }
+
+        wf::keyboard_focus_node_t keyboard_refocus(wf::output_t *output) override
+        {
+            wf::keyboard_focus_node_t node = {
+                .node = this,
+                .importance = wf::focus_importance::HIGH,
+                .allow_focus_below = false,
+            };
+            return node;
+        }
+
+        class lock_surface_keyboard_interaction : public wf::keyboard_interaction_t
+        {
+          public:
+            lock_surface_keyboard_interaction(wlr_surface *surface) : surface(surface)
+            {}
+
+            void handle_keyboard_enter(wf::seat_t *seat)
+            {
+                // TODO: use a keyboard grab instead?
```

Comment:
I think this is fine - wlroots grabs receive events if Wayfire decides to forward requests to wlroots - with HIGH focus importance and putting the lockscreen above everything else, Wayfire should always feed the focus to the lockscreen first.

Reply:
BTW, any reason not to use wf::input_manager_t::set_exclusive_focus here? Seems like it would be simpler. If the client crashes I assume that focus will go nowhere, but it sounds like that's what we want.

Reply:
set_exclusive_focus was meant for the older implementation of screenlockers, I think we can even remove this function nowadays. It is better to use the scenegraph API for this ;)